[MM-16379] Accessibility Navigation Controller

deanwhillier opened the pull request:
#### Summary
This PR adds in an accessibility keyboard navigation controller to globally control keyboard navigation around the webapp. This is a hybrid approach for an initial MVP but should probably be integrated into React/Redux in a future version. It provides 'hooks' into React for active/focused elements using custom event dispatching that can be listened for within appropriate React components (see below for details).

Initial container/element types supported include (classes are manually applied in appropriate React components):
* Regions (class = `a11y__region`): `(SHIFT+)CTRL+TILDE`, `(SHIFT+)F6` (Desktop), `(SHIFT+)CTRL+F6` (Browser)
   * Sections (class = `a11y__section`): `UP-ARROW`, `DOWN-ARROW`
      * Elements (no class): `(SHIFT+)TAB`

Region element attributes (manually applied in appropriate React components):
* `data-a11y-sort-order`: Define the navigation sort order for region elements using numbers (1-n) in ascending order or navigation preference
* `data-a11y-focus-child`: Focus immediately on first contained section of a given region instead of the region itself
* `data-a11y-order-reversed`: Reverse the order of navigation of contained sections

Regions and sections containing a focused element (or are themselves focused) will receive an `a11y--active` class. This can be used to style regions, sections and/or elements as needed when they are active.

Focused regions, sections and elements will receive an `a11y--focused` class (obviously only one at a time). This replaces the `keyboard-focus` class previously used for consistency.

Custom events are dispatched on region, section and element DOM elements when they become active and when they loose active status. These can be listened for in the corresponding React components (using ref) to do things as needed when they become active. For example, these events are used in the `Post` component to trigger creation/removal of the hover menu when a post is active during keyboard navigation. Events are as follows:
* `a11yactivate`: dispatched on DOM elements when they become active
* `a11ydeactivate`: dispatched on DOM elements when they loose active status

The controller listens for a custom event on all active/focused DOM elements, allowing the corresponding React components to inform to the controller when the component is updated. This allows the controller to maintain the `a11y--active` and `a11y--focused` classes. This needs to be setup in the component, see the Post component for an example. Event is as follows:
* `a11yupdate`: dispatched by React on DOM elements when active components update (`componentDidUpdate`)

Other:
* Tabbing in/out of a section/region will automatically update the active region/section, apply/remove the appropriate classes and dispatch the appropriate events (so long as the focused element is still within a region/section)
* Hitting escape at any point or clicking anywhere with the mouse will cancel keyboard navigation
* Navigating within a region will reset region navigation, meaning that the next region keyboard shortcut triggered will restart navigation at the first region
* The controller uses the capture phase for events it is listening for
* Regions and sections need to be visible to be included in navigation
* If a defined region and/or section is a non-focusable tag and doesn't have a `tabindex` attribute already set, one will be automatically applied and set to `-1` to allow it to be navigated to
* All a11y event, class and attribute names are available as constants

TODO:
* Build in support for modals/overlays and drop-downs.
* Select oldest unread post
* better scroll handling when changing focus to occluded elements

#### Ticket Link
[MM-16379](https://mattermost.atlassian.net/browse/MM-16379)

Example HTML:
``` html
<main>
    <section class="a11y__region" data-a11y-sort-order="1" data-a11y-focus-child="true">
        <div class="a11y__section">
            <ul>
                <li tabindex="0"></li>
                <li tabindex="0"></li>
            </ul>
        </div>
        <div class="a11y__section">
            <ul>
                <li tabindex="0"></li>
                <li tabindex="0"></li>
            </ul>
        </div>
        <div class="a11y__section">
            <ul>
                <li tabindex="0"></li>
                <li tabindex="0"></li>
            </ul>
        </div>
    </section>
    <section class="a11y__region" data-a11y-sort-order="3" data-a11y-order-reversed="true">
        <div class="a11y__section">
            <ul>
                <li tabindex="0"></li>
                <li tabindex="0"></li>
            </ul>
        </div>
        <div class="a11y__section">
            <ul>
                <li tabindex="0"></li>
                <li tabindex="0"></li>
            </ul>
        </div>
        <div class="a11y__section">
            <ul>
                <li tabindex="0"></li>
                <li tabindex="0"></li>
            </ul>
        </div>
    </section>
    <section class="a11y__region" data-a11y-sort-order="2">
        <div class="a11y__section">
            <ul>
                <li tabindex="0"></li>
                <li tabindex="0"></li>
            </ul>
        </div>
        <div class="a11y__section">
            <ul>
                <li tabindex="0"></li>
                <li tabindex="0"></li>
            </ul>
        </div>
        <div class="a11y__section">
            <ul>
                <li tabindex="0"></li>
                <li tabindex="0"></li>
            </ul>
        </div>
    </section>
</main>
```

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-02038d3fe9cbbe50b.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.27.204
Instance ID: i-02038d3fe9cbbe50b

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0aa726518f3edf5f1.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.15.220
Instance ID: i-0aa726518f3edf5f1

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0558c61b0667c1999.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.59.255
Instance ID: i-0558c61b0667c1999

deanwhillier commented:
Updated the description with some additional details. Will deal with the conflicting files now.

deanwhillier commented:
Hey @esethna, thanks for the review!

First some general comments:
* I have not had a chance to test the controller again since merging master this morning to perform necessary conversions between @asaadmahmood's work and mine, as I'm currently sequestered on a couple urgent tickets. I'm guessing this is the reason for #3, #6
* This PR is for the navigation controller itself (navigating between regions, sections and generally keeping track of focus while keyboard navigation is active) and not for specifically ensuring that tab order and visual focus are correct yet. This is likely the reason for #1, #4, #5, #6.
* I'll be working on specific post tabbing order, visual indicators etc. in [MM-16380](https://mattermost.atlassian.net/browse/MM-16380)

More specific comments:
* #1: will be addressed as a part of [MM-16380](https://mattermost.atlassian.net/browse/MM-16380)
* #2: the controller is currently only engaged when starting with F6 but can also engage when any element receives focus by tabbing – I can update it to function this way instead. I just wasn't sure if any tab at any time should engage keyboard navigation functionality/styling or not as part of this effort.
* #3: Likely a merge related challenge as mentioned above
* #4: The page structure doesn't readily support this though general tabbing as the Search bar, `@` button and Flag button are duplicated in the header and the RHS.  When the RHS is hidden, the elements in the header are shown, when the RHS is open, the elements in the RHS are shown, messing up tabbing order in the channel header. Some specific overrides will need to be added as part of a ticket for handling channel header keyboard navigation. I can likely accommodate this with additional controller functionality (I have some thoughts on this already).
* #5: will be addressed as a part of [MM-16380](https://mattermost.atlassian.net/browse/MM-16380)
* #6: Likely a merge related challenge as mentioned above, but will specifically be addressed as a part of [MM-16380](https://mattermost.atlassian.net/browse/MM-16380)
* #7: Not a problem to handle with the controller, I just need to tweak it to support.

enahum commented on components/rhs_comment/rhs_comment.jsx:
```diff
@@ -382,10 +409,12 @@ export default class RhsComment extends React.PureComponent {
-                ref={'post_body_' + post.id}
+                ref={this.postRef}
+
+                // ref={'post_body_' + post.id}
```
nit: comment line not needed

enahum commented on components/post_view/post/post.jsx:
```diff
@@ -252,10 +275,10 @@ export default class Post extends React.PureComponent {
                 id={'post_' + post.id}
                 role='listitem'
-                className={this.getClassName(post, isSystemMessage, isMeMessage, fromWebhook, fromAutoResponder, fromBot)}
+                className={'a11y__section ' + this.getClassName(post, isSystemMessage, isMeMessage, fromWebhook, fromAutoResponder, fromBot)}
```
```suggestion
                className={`a11y__section ${this.getClassName(post, isSystemMessage, isMeMessage, fromWebhook, fromAutoResponder, fromBot)}`}
```

enahum commented on components/rhs_comment/rhs_comment.jsx:
```diff
@@ -382,10 +409,12 @@ export default class RhsComment extends React.PureComponent {
                 id={'rhsPost_' + post.id}
                 tabIndex='-1'
-                className={this.getClassName(post, isSystemMessage)}
+                className={'a11y__section ' + this.getClassName(post, isSystemMessage)}
```
```suggestion
                className={`a11y__section ${this.getClassName(post, isSystemMessage)}`}
```

enahum commented on components/rhs_root_post/rhs_root_post.jsx:
```diff
@@ -308,7 +308,7 @@ export default class RhsRootPost extends React.PureComponent {
                 id={'rhsPost_' + post.id}
                 tabIndex='-1'
-                className={'thread__root ' + this.getClassName(post, isSystemMessage)}
+                className={'thread__root a11y__section ' + this.getClassName(post, isSystemMessage)}
```
```suggestion
                className={`thread__root a11y__section ${this.getClassName(post, isSystemMessage)}`}
```

deanwhillier commented:
1. Buttons and hover menu should now stay. There is still an issue with the flag component that needs to be addressed (not related to the controller) and the rest of the functionality for the hover menu will be handled in another ticket.
2. You can now tab from anywhere and the controller will automatically determine what region and/or section you are in to allow for proper section navigation (up/down) and the focused elements will receive the correct class names to apply the proper styling.
3. Turned keyboard tabbing on (was disabled previously) and made some tweaks to the controller to allow regions and sections to also be focused elements.
4. Not solved yet, will need to be a separate ticket. This one will be a bit more challenging due to the duplicate search, @ and flag buttons in the actual header and the RHS. The structural order of the markup is what is messing up the tabbing order. Still have some thoughts, but will need to be solved separately.
5. Similar to #1, this needs to be solved at the component level separately to this PR
7. Up/Down section navigation will now loop in bother directions

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-073ae7846054d46ed.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.41.158
Instance ID: i-073ae7846054d46ed

deanwhillier commented:
There is currently an interesting click effect that I'm troubleshooting. An artifact of being able to click anywhere and hit tab to engage keyboard navigation.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-02425d30f9c7b3145.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.38.154
Instance ID: i-02425d30f9c7b3145

mattermod commented:
Spinmint test server destroyed

deanwhillier commented:
Master has been merged. There will likely be some tweaks needed, but a cursory test seems to be working for me.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0bb9dc35332ded882.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.58.254
Instance ID: i-0bb9dc35332ded882

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Spintmint test server creation failed. See the logs for more information.

mattermod commented:
Spintmint test server creation failed. See the logs for more information.

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-3047.test.mattermost.cloud

Test Admin Account: Username: `sysadmin` | Password: `Sys@dmin123`

Test User Account: Username: `user-1` | Password: `User-1@123`

mattermod commented:
Spinmint test server destroyed

deanwhillier commented:
> Minor issue, can be filed as a separate ticket if this is difficult: In the post list, hitting down arrow on the newest post **should not** loop to the first post in the channel. This is a pretty bad experience given you could potentially be moved to a message way off screen unexpectedly.

@esethna yes, I noticed that as well when I switched to up/down arrow key looping, meant to bring it up. I added an attribute override to manually change the default looping state within a region. Applied it to just the posts in the centre so no more looping. The other regions still loop as needed.

esethna commented:
Removing EN request for review, he said we can merge
